File: python/2406. Divide Intervals Into Minimum Number of Groups.py
class Solution:
    def minGroups(self, intervals: List[List[int]]) -> int:
        intervals.sort()

        max_groups = 0
        heap = []

        for (beg, end) in intervals:
            while heap and heap[0] < beg:
                heapq.heappop(heap)
            heapq.heappush(heap, end)
            max_groups = max(max_groups, len(heap))
        return max_groups

File: rust/2406. Divide Intervals Into Minimum Number of Groups.rs
use std::cmp::{max, Reverse};
use std::collections::BinaryHeap;

impl Solution {
    pub fn min_groups(mut intervals: Vec<Vec<i32>>) -> i32 {
        intervals.sort_by(|a, b| a[0].cmp(&b[0]));

        let mut heap = BinaryHeap::new();
        let mut max_groups = 0;

        for interval in intervals.iter() {
            while let Some(Reverse(top)) = heap.peek() {
                if top < &interval[0] {
                    heap.pop();
                } else {
                    break;
                }
            }
            heap.push(Reverse(interval[1]));
            max_groups = std::cmp::max(max_groups, heap.len());
        }

        max_groups as i32
    }
}
